### PR TITLE
3579 entity generated id property

### LINKED ISSUE #3579: Not-Unique ID field (20 pts)
Create an ID field that will automatically generate a code for each entity.

- The code should be 2 letters and 4 numbers (discuss collision in big collections) (sequential?)
- The client should pre-fill the field with a generated code
- In the scenarion of the public submit, is the backend who will generate the code.
- The code should be editable
- The code may not be unique in the DB
-  We support multiple blank values

property in template builder

To consider:
 - On public form, this should not be a field that gets rendered, but assigned a value upon creation. Perhaps zero-out on the public endpoint and then auto-fill on backend if field is missing? (discuss)
 - Copy from should not copy this value
 - CSV import should generate this value if not present (probably done by the entity save process)

Designs: https://www.figma.com/file/xjsz7r65epN4B0b96zilac/0-uwazi-redesigns

Long discussion #1173

### PR DESCRIPTION
**Pull request by mfacar**
fixes #3579

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

### CODE REVIEW

**Review by fnocetti — CHANGES_REQUESTED** [review]
@mfacar if this approach is the correct, then it LGTM but for a couple of comments. Let's present it to the team.
I still need to test this, on the while, please, take a look :)

File: app/react/Metadata/actions/actions.js [hunk 1]
```diff
@@ -33,17 +54,9 @@ export const resetMetadata = (metadata, template, options, previousTemplate) =>
...
-      resetedMetadata[name] = { from: null, to: null };
-    }
-    if (
-      resetValue &&
-      ['multiselect', 'relationship', 'nested', 'multidate', 'multidaterange'].includes(type)
-    ) {
-      resetedMetadata[name] = [];
+    if (resetValue) {
+      const defaultValue = defaultValueByType(type, options);
+      if (defaultValue !== undefined) resetedMetadata[name] = defaultValue;
```
[the remark this answers is not in the document]
Comment (mfacar) [reply app/react/Metadata/actions/actions.js:59]: No, it is not enough, there are some types that have as a valid default value an empty string.

File: app/shared/IDGenerator.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+const limitedLength = (length: number) => Math.max(0, Math.min(13, length));
+const IDGenerator = {
```
[the remark this answers is not in the document]
Comment (mfacar) [reply app/shared/IDGenerator.ts:2]: No reason, I'll change it

File: app/react/Metadata/actions/actions.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (fnocetti) [reply app/react/Metadata/actions/actions.js:59]: Gotcha!

**Comment by fnocetti** [comment]
@mfacar The functionality looks good en general. There's two bugs:

- When using the ID field as a filter it doesn't work. It's not filtering and also shows [object Object] as in the image bellow when blank.

Screen Shot 2021-04-29 at 5 42 26 PM

- CSV import is not generating the value. Used the following csv on the Document template (default):

`title,generated id
con, GENID
sin,`

The first one is taking the provided value while the second is staying empty.

Please check. Thanks!

**Comment by mfacar** [comment]
Good catches, thank you Fede, I'm working on them

**Comment by mfacar** [comment]
@fnocetti, the bugs are fixed, thanks for reporting them. I found a problem at the searching by this type of property which is also fixed. Please take a look.

**Review by fnocetti — APPROVED** [review]
Looks good! Good job! Merging now

**Pull request merged**